HPCC-25360 Convert sasha service list to a dictionary

jakesmith opened the pull request:
Also fix an issue if secrets configured, and multiple dali
sasha services used them and generated the same volume>1

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith committed:
HPCC-25360 Convert sasha service list to a dictionary

Also fix an issue if secrets configured, and multiple dali
sasha services used them and generated the same volume>1

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-14583, label: hpcc-25360-sasha-dict is skipped.
The reason is: It is related to containerised environment only.
Commit ID: 417c655027ec6b76a972fda05642f1beb8a581e2
OS: centos 7.7.1908 (Linux 3.10.0-327.10.1.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: hpcc-platform-dev-el7-smoketest-2

shamser commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -609,17 +609,19 @@ Generate list of available services
 {{- range $dali := $.Values.dali -}}
-{{- range $sasha := $dali.services -}}
+{{- range $sashaName, $_sasha := $dali.services -}}
+{{- $sasha := ($_sasha | default dict) -}}
```
What's the purpose of 'default dict' here?

shamser commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -609,17 +609,19 @@ Generate list of available services
 {{ end -}}
-{{- range $sasha := $.Values.sasha -}}
+{{- range $sashaName, $_sasha := $.Values.sasha -}}
+{{- $sasha := ($_sasha | default dict) -}}
```
Again, is 'default dict' needed here?

shamser commented on helm/hpcc/values.schema.json:
```diff
@@ -799,70 +798,81 @@
+            "at" : {},
+            "throttle": {},
+            "storage": {},
+            "resources": {}
```
The attributes above do not have a type or description

shamser replied:
A lot of fields here.  Could we simplify values.yaml by defining defaults for most of these?

jakesmith replied:
actually, they are still needed.
The schemas are defined and inherited via allOf and "sashacommon".
They are necessary to define here too, because the additionalProperties: false only applies to "properties", so although the schema are inherited, they need defining here to avoid additionalProperties rejecting them when they appear in values.yaml.
For reference see: https://json-schema.org/understanding-json-schema/reference/combining.html.

However, I did notice a mistake in the use of allOf, I'll change in coming commit

shamser commented on helm/hpcc/values.schema.json:
```diff
@@ -799,70 +798,81 @@
-      },
-      "additionalProperties": false
+        {
+          "type": "null"
```
what's the purpose of this?

shamser commented on helm/hpcc/values.schema.json:
```diff
@@ -887,63 +897,87 @@
+            "storage": {},
+            "resources": {},
+            "limit": {},
+            "cutoff": {}
```
The attributes above do not have a type or description

shamser commented on helm/hpcc/values.yaml:
```diff
@@ -234,70 +234,70 @@ bundles: []
+      #interval: 2 # (hours)
+      #at: "* * * * *" # cron type schedule, i.e. Min(0-59) Hour(0-23) DayOfMonth(1-31) Month(1-12) DayOfWeek(0-6)
+      #minDeltaSize: 50 # (Kb) will not start coalescing until delta log is above this threshold
+      #recoverFromIncErrors: false # Try to continue after transaction load failures
```
recoverFromIncErrors is not defined in the schema.

jakesmith replied:
I didn't mean for this to be here, will remove.

shamser reviewed (commented):
@jakesmith

jakesmith committed:
HPCC-25360 Fix a couple of schema issues

1) Fix use of allOf with additioanl properties
2) Remove unwanted values.yaml commented reference to
recoverFromIncErrors
3) Remove superfluous persistExpiryDefault and expiryDefault from
dfuwu-archiver
4) Add missing "oneOf" in sasha-dfuwu-archiver schema

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented:
@shamser - please review new commit

HPCCSmoketest commented:
Process of PR-14583, label: hpcc-25360-sasha-dict is skipped.
The reason is: It is related to containerised environment only.
Commit ID: b2456f08ea458a34693754eebcaf1c3d7d3a8fda
OS: centos 7.7.1908 (Linux 3.10.0-327.10.1.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: hpcc-platform-dev-el7-smoketest-2

shamser reviewed (approved):
@richardkchapman Reviewed.
